cryptol-remote-api/src/CryptolServer/Sat.hs

```diff
@@ -27,76 +32,162 @@ import Cryptol.ModuleSystem (checkExpr)
-              AllSatResult results ->
-                Satisfied <$> traverse satResult results
+proveSat (ProveSatParams queryType (ProverName proverName) jsonExpr hConsing) = do
+  e <- getExpr jsonExpr
+  (_expr, ty, schema) <- liftModuleCmd (checkExpr e)
+  -- TODO validEvalContext expr, ty, schema
+  decls <- deDecls . meDynEnv <$> getModuleEnv
+  s <- getTCSolver
+  perhapsDef <- liftIO (defaultReplExpr s ty schema)
+  case perhapsDef of
+    Nothing -> raise (evalPolyErr schema)
+    Just (_tys, checked) -> do
+      timing <- liftIO $ newIORef 0
+      let cmd =
+            ProverCommand
+            { pcQueryType    = queryType
+            , pcProverName   = proverName
+            , pcVerbose      = True
+            , pcProverStats  = timing
+            , pcExtraDecls   = decls
+            , pcSmtFile      = Nothing
+            , pcExpr         = checked
+            , pcSchema       = schema
+            , pcValidate     = False
+            , pcIgnoreSafety = False
+            }
+      res <- if proverName `elem` ["offline","sbv-offline","w4-offline"]
+             then offlineProveSat proverName cmd hConsing
+             else onlineProveSat  proverName cmd hConsing
+      _stats <- liftIO (readIORef timing)
+      pure res
+
+
+getProverConfig ::
+  -- | Prover name.
+  String ->
+  CryptolCommand (Either SBV.SBVProverConfig W4.W4ProverConfig)
+getProverConfig proverName =
+  if proverName `elem` W4.proverNames then do
+    liftIO (W4.setupProver proverName) >>= \case
+      Left msg -> raise $ proverError $ "error setting up " ++ proverName ++ ": " ++ msg
+      Right (_ws, w4Cfg) -> pure $ Right w4Cfg
+  else if proverName `elem` SBV.proverNames  then do
+    liftIO (SBV.setupProver proverName) >>= \case
+      Left msg -> raise $ proverError $ "error setting up " ++ proverName ++ ": " ++ msg
+      Right (_ws, sbvCfg) -> pure $ Left sbvCfg
+  else do
+    raise $ proverError
+          $ proverName ++ "is not a recognized solver;"
+            ++ " please choose one of the following instead: "
+            ++ (show $ W4.proverNames ++ SBV.proverNames)
 
+
+offlineProveSat ::
+  -- | Prover name.
+  String ->
+  ProverCommand ->
+  -- | Whether hash consing is enabled.
+  Bool ->
+  CryptolCommand ProveSatResult
+offlineProveSat proverName cmd hConsing = do
+  getProverConfig proverName >>= \case
+    Left sbvCfg -> do
+      result <- liftModuleCmd $ SBV.satProveOffline sbvCfg cmd
+      case result of
+        Left msg -> do
+          raise $ proverError $ "error setting up " ++ proverName ++ ": " ++ msg
+        Right smtlib -> pure $ OfflineSMTQuery smtlib
+    Right w4Cfg -> do
+      smtlibRef <- liftIO $ newIORef ""
+      result <- liftModuleCmd $
+        W4.satProveOffline w4Cfg hConsing False cmd $ \f -> do
+          withRWTempFile "smtOutput.tmp" $ \h -> do
+            f h
+            hSeek h AbsoluteSeek 0
+            contents <- TIO.hGetContents h
+            writeIORef smtlibRef $ T.unpack contents
```

Comment:
Maybe don't unpack and make the field of `OfflineSMTQuery` a `Text` value instead? The contents of these files can get large sometimes.